## Data sources

### Toronto Traffic Collisions
- **Source:** Toronto Police Service Public Safety Data Portal — Traffic Collisions Open Data (ASR-T-TBL-001)
- **Downloaded:** 2026-10-05
- **Format:** CSV
- **Coverage in downloaded snapshot:** 2014-01-01 — 2026-06-30
- **Licence:** Open Government Licence

### KSI Collisions
- **Source:** City of Toronto Open Data — Motor Vehicle Collisions with Killed or Seriously Injured (KSI) Data
- **Downloaded:** 2026-10-05
- **Resource:** Motor Vehicle Collisions with KSI Data
- **Format:** CSV, WGS84
- **Coverage in downloaded snapshot:** 2006-01-01 — 2026-09-15
- **Licence:** Open Government Licence

### Usage note
The datasets are used for research and portfolio development. Source attribution must be preserved and the applicable Open Government Licence and portal terms must be respected.

In [36]:
import pandas as pd

df = pd.read_csv("../data/raw/Traffic_Collisions_Open_Data_8128730402587031536.csv")

df.head()

,OBJECTID,EVENT_UNIQUE_ID,OCC_DATE,OCC_MONTH,OCC_DOW,OCC_YEAR,OCC_HOUR,DIVISION,FATALITIES,INJURY_COLLISIONS,...,NEIGHBOURHOOD_158,LONG_WGS84,LAT_WGS84,AUTOMOBILE,MOTORCYCLE,PASSENGER,BICYCLE,PEDESTRIAN,x,y
0,1,GO-20148000059,1/1/2014 5:00:00 AM,January,Wednesday,2014,15,D43,0,NO,...,Centennial Scarborough (133),-79.172461,43.782185,YES,NO,NO,NO,NO,-8.813438e+06,5.431796e+06
1,2,GO-20148049789,1/1/2014 5:00:00 AM,January,Wednesday,2014,11,D32,0,NO,...,Englemount-Lawrence (32),-79.432440,43.720468,YES,NO,NO,NO,NO,-8.842379e+06,5.422285e+06
2,3,GO-20141260499,1/1/2014 5:00:00 AM,January,Wednesday,2014,2,NSA,0,YES,...,NSA,0.000000,0.000000,YES,NO,NO,NO,NO,6.327780e-09,5.664924e-09
3,4,GO-20141260663,1/1/2014 5:00:00 AM,January,Wednesday,2014,3,NSA,0,NO,...,NSA,0.000000,0.000000,YES,NO,NO,NO,NO,6.327780e-09,5.664924e-09
4,5,GO-20141261162,1/1/2014 5:00:00 AM,January,Wednesday,2014,5,NSA,0,YES,...,NSA,0.000000,0.000000,YES,NO,NO,NO,NO,6.327780e-09,5.664924e-09


In [37]:
ksi_df = pd.read_csv("../data/raw/Motor Vehicle Collisions with KSI Data - 4326.csv")

ksi_df.head()

,_id,collision_id,accdate,stname1,stname2,stname3,per_inv,acclass,accloc,traffictl,...,distracted,cyclist,motorcyclist,other_micromobility,older_adult,pedestrian,red_light,school_child,heavy_truck,geometry
0,1,2006:893184,2006-01-01T02:36:00,WOODBINE AVE,O CONNOR DR,NaN,8,Non-Fatal Injury,Intersection-Related,No Control,...,False,False,False,False,False,False,False,True,False,"{""coordinates"": [[-79.31879768423914, 43.69959..."
1,2,2006:893184,2006-01-01T02:36:00,WOODBINE AVE,O CONNOR DR,NaN,8,Non-Fatal Injury,Intersection-Related,No Control,...,False,False,False,False,False,False,False,True,False,"{""coordinates"": [[-79.31879768423914, 43.69959..."
2,3,2006:893184,2006-01-01T02:36:00,WOODBINE AVE,O CONNOR DR,NaN,8,Non-Fatal Injury,Intersection-Related,No Control,...,False,False,False,False,False,False,False,True,False,"{""coordinates"": [[-79.31879768423914, 43.69959..."
3,4,2006:893184,2006-01-01T02:36:00,WOODBINE AVE,O CONNOR DR,NaN,8,Non-Fatal Injury,Intersection-Related,No Control,...,False,False,False,False,False,False,False,True,False,"{""coordinates"": [[-79.31879768423914, 43.69959..."
4,5,2006:893184,2006-01-01T02:36:00,WOODBINE AVE,O CONNOR DR,NaN,8,Non-Fatal Injury,Intersection-Related,No Control,...,False,False,False,False,False,False,False,True,False,"{""coordinates"": [[-79.31879768423914, 43.69959..."


In [ ]:
print(
    "сколько в таблице строк и столбцов", df.shape,
    "полный список названий колонок", df.columns,
    "типы данных колонок", df.dtypes, sep='\n'
    )

сколько в таблице строк и столбцов
(825265, 23)
полный список названий колонок
Index(['OBJECTID', 'EVENT_UNIQUE_ID', 'OCC_DATE', 'OCC_MONTH', 'OCC_DOW',
       'OCC_YEAR', 'OCC_HOUR', 'DIVISION', 'FATALITIES', 'INJURY_COLLISIONS',
       'FTR_COLLISIONS', 'PD_COLLISIONS', 'HOOD_158', 'NEIGHBOURHOOD_158',
       'LONG_WGS84', 'LAT_WGS84', 'AUTOMOBILE', 'MOTORCYCLE', 'PASSENGER',
       'BICYCLE', 'PEDESTRIAN', 'x', 'y'],
      dtype='str')
типы данных колонок
OBJECTID               int64
EVENT_UNIQUE_ID          str
OCC_DATE                 str
OCC_MONTH                str
OCC_DOW                  str
OCC_YEAR               int64
OCC_HOUR               int64
DIVISION                 str
FATALITIES             int64
INJURY_COLLISIONS        str
FTR_COLLISIONS           str
PD_COLLISIONS            str
HOOD_158                 str
NEIGHBOURHOOD_158        str
LONG_WGS84           float64
LAT_WGS84            float64
AUTOMOBILE               str
MOTORCYCLE               str
PASSENGER      

In [39]:
print(
    "сколько в таблице строк и столбцов ksi", ksi_df.shape,
    "полный список названий колонок ksi", ksi_df.columns,
    "типы данных колонок ksi", ksi_df.dtypes, sep='\n'
    )

сколько в таблице строк и столбцов ksi
(20771, 50)
полный список названий колонок ksi
Index(['_id', 'collision_id', 'accdate', 'stname1', 'stname2', 'stname3',
       'per_inv', 'acclass', 'accloc', 'traffictl', 'impactype', 'visible',
       'light', 'rdsfcond', 'road_class', 'failtorem', 'longitude', 'latitude',
       'veh_no', 'vehtype', 'initdir', 'per_no', 'invage', 'injury',
       'safequip', 'drivact', 'drivcond', 'pedact', 'pedcond', 'manoeuvre',
       'pedtype', 'cyclistype', 'cycact', 'cyccond', 'road_user', 'fatal_no',
       'wardname', 'division', 'neighbourhood', 'aggressive', 'distracted',
       'cyclist', 'motorcyclist', 'other_micromobility', 'older_adult',
       'pedestrian', 'red_light', 'school_child', 'heavy_truck', 'geometry'],
      dtype='str')
типы данных колонок ksi
_id                      int64
collision_id               str
accdate                    str
stname1                    str
stname2                    str
stname3                    str
per_in

In [ ]:
print("Число уникальных ДТП:", df.EVENT_UNIQUE_ID.nunique())
print("Число повторов:", df.EVENT_UNIQUE_ID.duplicated().sum())

Число уникальных ДТП: 825265
Число повторов: 0


In [41]:
print("Число уникальных ДТП ksi:", ksi_df.collision_id.nunique())
print("Число повторов ksi:", ksi_df.collision_id.duplicated().sum())

Число уникальных ДТП ksi: 7620
Число повторов ksi: 13151


In [42]:
df.OCC_DATE = pd.to_datetime(df.OCC_DATE)
print("Временной промежуток в DF:", df.OCC_DATE.min(), "-", df.OCC_DATE.max())

Временной промежуток в DF: 2014-01-01 05:00:00 - 2026-06-30 04:00:00


In [43]:
mask = ksi_df.collision_id == "2006:893184"

ksi_df.loc[mask, ["collision_id", "injury", "road_user", "per_no"]]

,collision_id,injury,road_user,per_no
0,2006:893184,Minor,passenger,4.0
1,2006:893184,Minor,driver,2.0
2,2006:893184,Minor,passenger,3.0
3,2006:893184,Major,passenger,6.0
4,2006:893184,Minor,passenger,1.0
5,2006:893184,Minor,passenger,5.0
6,2006:893184,NaN,driver,8.0
7,2006:893184,NaN,other,7.0


In [31]:
print(df.INJURY_COLLISIONS.value_counts())
print("\n", df.FATALITIES.value_counts())

INJURY_COLLISIONS
NO     712967
YES    112294
Name: count, dtype: int64

 FATALITIES
0    824582
1       673
2         6
3         3
4         1
Name: count, dtype: int64


In [44]:
ksi_df.accloc.value_counts(dropna=False)

accloc
At Intersection          8877
NaN                      5456
Non-Intersection         3332
Intersection-Related     2206
At/Near Private Drive     514
Four-way                  227
T Intersection             64
Laneway                    22
Private Driveway           15
Overpass or Bridge         15
Other Intersection         12
Underpass or Tunnel         9
Y Intersection              6
Roundabout                  4
Offset                      4
L Intersection              4
At Railway Crossing         2
Trail                       2
Name: count, dtype: int64

In [45]:
ksi_df.traffictl.value_counts(dropna=False)

traffictl
No Control              9650
Traffic Signal          8828
Stop Sign               1601
Pedestrian Crossover     250
NaN                      177
Traffic Controller       110
Unknown                  102
Yield Sign                22
Streetcar (Stop for)      16
Traffic Gate               5
Other                      2
Half Signal                2
Pedestrian Signal          2
Police Control             2
School Guard               2
Name: count, dtype: int64

In [48]:
ksi_df.loc[ksi_df.collision_id == "2006:893184", "traffictl"].unique()

<StringArray>
['No Control']
Length: 1, dtype: str

In [49]:
ksi_df.loc[ksi_df.collision_id == "2006:893184", "accloc"].unique()

<StringArray>
['Intersection-Related']
Length: 1, dtype: str

In [51]:
ksi_df.groupby("collision_id")["traffictl"].nunique(dropna=False).value_counts()

traffictl
1    7620
Name: count, dtype: int64

In [52]:
ksi_df.groupby("collision_id")["accloc"].nunique(dropna=False).value_counts()

accloc
1    7620
Name: count, dtype: int64

In [54]:
ksi_events = ksi_df.drop_duplicates(subset="collision_id")
ksi_events.shape

(7620, 50)

In [59]:
ksi_events.accloc.isna().sum()

np.int64(1841)

In [57]:
ksi_events.traffictl.isna().sum()

np.int64(75)

In [58]:
ksi_events.longitude.isna().sum()

np.int64(1)

In [60]:
ksi_events.latitude.isna().sum()

np.int64(1)

In [62]:
print(
    (ksi_events.longitude == 0).sum(),
    (ksi_events.latitude == 0).sum(), sep='\n'
)

0
0


In [63]:
ksi_events[["longitude", "latitude"]].isna().any(axis=1).sum()

np.int64(1)

In [64]:
(ksi_events.traffictl == "Traffic Signal").sum()

np.int64(3107)

In [65]:
mask = ksi_events.traffictl == "Traffic Signal"

ksi_events.loc[mask, "accloc"].value_counts(dropna=False)

accloc
At Intersection          2151
Intersection-Related      480
NaN                       309
Four-way                   80
Non-Intersection           57
T Intersection             16
At/Near Private Drive       8
Other Intersection          4
Roundabout                  1
Offset                      1
Name: count, dtype: int64

In [67]:
intersection_categories = [
    "At Intersection",
    "Four-way",
    "T Intersection",
    "Other Intersection",
    "Roundabout"
]

ksi_events.accloc.isin(intersection_categories).value_counts()

accloc
False    4329
True     3291
Name: count, dtype: int64

In [68]:
strict_mask = (
    (ksi_events["traffictl"] == "Traffic Signal")
    &
    (ksi_events["accloc"].isin(intersection_categories))
)

strict_mask.sum()

np.int64(2252)

In [71]:
ksi_df.accdate = pd.to_datetime(ksi_df.accdate)
print("Временной промежуток в DF ksi:", ksi_df.accdate.min(), "-", ksi_df.accdate.max())

Временной промежуток в DF ksi: 2006-01-01 02:36:00 - 2026-09-15 13:16:00


## Feasibility verdict

**Decision: GO**

The Toronto pilot is feasible with the current open data.

- The general Traffic Collisions dataset contains 825,265 unique collision events and is suitable for overall collision context, but it does not distinguish serious injuries from other injuries.
- The actively maintained City of Toronto KSI dataset provides the required KSI definition and contains 20,771 person-level records representing 7,620 unique collision events.
- Event-level road context is internally consistent for `traffictl` and `accloc`.
- 7,619 of 7,620 KSI collision events have complete latitude and longitude.
- Using the strict pilot definition (`Traffic Signal` + explicit intersection location categories), 2,252 unique KSI collision events are available.

**Constraint:** the KSI dataset, not the general Traffic Collisions dataset, should be used as the source of the KSI outcome. `Intersection-Related` is excluded from the strict pilot definition until its interpretation is justified more clearly.